### ISSUE TITLE
html formatted text boxes do not create links in pdf outputs

### ISSUE THREAD

**Issue by qgib** [qgis/QGIS#19117]
Author Name: **Giovanni Manghi** (@gioman)
Original Redmine Issue: [10731](https://issues.qgis.org/issues/10731)

Redmine category:map_composer/printing

---
Not sure if this is a feature request or a bug,

if I create a text box in a composer and add a tag

<a href="http://www.google.com/">link</a>

when outputting to PDF the link text is formatted as expected (blue, underlined), but the link does not work (maybe it does not exist as it is not clickable).

It does not depend on the platform (tested Linux and Windows) and also does not depend on how the PDF is created (the QGIS function or other virtual printer).

**Comment by suricactus** [comment]
5.12 and current master still have the same issue, QT 5.12.4.

**Comment by ThomasG77** [comment]
Not sure how it works on QGIS side but I'm able with pure PyQt to export PDF with link using a `QPrinter` and `QPainter` e.g https://gist.github.com/ThomasG77/72fd711bda2ec0c8b95a0dacd2dbb942 (QGIS build with Qt 5.9.5)

From what I've seen so far, `QgsLayoutExporter::exportToPdf` (https://qgis.org/api/qgslayoutexporter_8cpp_source.html#l00507) calls `QgsLayoutExporter::printPrivate` (https://qgis.org/api/qgslayoutexporter_8cpp_source.html#l01206) with `QPainter` and `QPrinter` among the args.
This function calls `QgsLayoutExporter::renderPage` (https://qgis.org/api/qgslayoutexporter_8cpp_source.html#l00154) that call `QgsLayoutExporter::renderRegion` (https://qgis.org/api/qgslayoutexporter_8cpp_source.html#l00244). It calls a the painter as an arg and then make a call to render QgsLayout to painter. Although I've started tracking in the code how PDF rendering happens, I'm always unable to confirm where exactly the issue happens. Need more investigation. Looking for others input.

**Comment by signedav** [comment]
@ThomasG77 Have you done any more investigations, or is this the status to pick up?

**Comment by ThomasG77** [comment]
@signedav Did not have time to investigate more. After all, not a low level dev, difficult for me to go further.

**Comment by m-kuhn** [comment]
Some further analysis on this.

The Qt printing API does not allow generating links through public API.

What existing Qt classes use to deal with this:

- `QTextDocument` uses a purely internal, private implementation (`QTextItemInt`) which is understood by the pdf exporter code
- `QWebView` uses a dedicated printing function which very likely cannot easily be mixed with other generic painting code (like `QGraphicsScene`)

A first proof of concept shows, that with `QTextDocument` it is possible to add links within an ongoing printing process by directly using the layout.draw() function.

Implementing this in QGIS should be possible.
The limitations of this are:
1. Only a subset of HTML is supported: https://doc.qt.io/qt-5/richtext-html-subset.html . It should therefore be added as a new option for HTML items (export labels at the expense of supporting a limited set of HTML)
2. The link is also clickable if it is outside the clipped region. Might be solvable.

<details>
<summary>Sample code that mixes <code>QGraphicsScene</code> and <code>QTextDocument</code></summary>
```
from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from PyQt5.QtPrintSupport import *
from PyQt5.Qt import *

app = QApplication([])
doc = QTextDocument()
doc.setHtml('<a href="http://qfield.org/get">go get it</a>')
layout = doc.documentLayout()

printer = QPrinter()
printer.setOutputFormat(QPrinter.PdfFormat)
printer.setOutputFileName("/tmp/output.pdf")

scene = QGraphicsScene()
scene.addRect(100, 0, 20, 100, QPen(Qt.black), QBrush(Qt.green))

painter = QPainter(printer)
scene.render(painter)
ctx = QAbstractTextDocumentLayout.PaintContext()
painter.save()
painter.setClipRect(QRectF(50, 50, 100, 100))
painter.translate(QPoint(100, 100))
layout.draw(painter, ctx)
painter.restore()
painter.end()


print(dir(layout))
```
</details>

**Comment by nirvn** [comment]
Unless I'm missing part of the intent of the author who filed this issue, this has been implemented in QGIS already.

**Comment by ThomasG77** [comment]
Not really except if I'm using an old QGIS version. See the project attachment (within zip). Export to pdf the layout (also attached [Mise en page 1.pdf](https://github.com/qgis/QGIS/files/14068520/Mise.en.page.1.pdf)) and see that the content although blue does not behave as a link. You can click as much as you want: nothing happen or at least for me.
[debug_export_pdf.zip](https://github.com/qgis/QGIS/files/14068539/debug_export_pdf.zip)

**Comment by signedav** [comment]
What version you are using @ThomasG77 afaik it should work since 3.32

**Comment by ThomasG77** [comment]
Don't ask... Too old 3.28. If it works on more recent versions and you confirm, close as suggested by @nirvn (my error here if fixed but only in recent versions)

**Comment by nirvn** [comment]
Functionality implemented in 3.32: https://www.qgis.org/en/site/forusers/visualchangelog332/index.html#feature-hyperlink-support-for-html-enabled-label-exports
